ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +188,82 @@ public void shutdown (boolean async)
-     * progress when this method is called it will complete its scan without suspending.
-     *
-     * @param async false if the calling thread should wait for any in-progress scan to complete before returning
-     * @return the previous mode before attempting the suspend
+     * <p>
+     *     Make all scanning operations suspend.
+     * </p>
+     * <p>
+     *     This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
+     *     threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
+     *     progress when this method is called it will complete its scan without suspending.
+     * </p>
+     * <p>
+     *     Note that this method is also influenced by
+     *     <code>RecoveryEnvironmentBean.setWaitForFinalRecovery(boolean)</code>.
+     *     Please, refer to the its javadoc for further details.
+     *     In case <code>isWaitForFinalRecovery() == true</code>, it is <b>very important</b> that, before invoking
+     *     <code>suspendScan</code>:
+     *     * The Transaction System (i.e. TxControl) gets disabled (i.e. no new transactions will be created)
+     *     * The Transaction Reaper finishes monitoring all in-flight transactions
+     * </p>
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     *              In case <code>isWaitForFinalRecovery() == true</code>, this parameter is override.
+     * @return the previous mode before attempting the suspension
      */
 
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForFinalRecovery checks whether suspension should delay until RecoveryModules
+            * implementing SuspendBlockingRecoveryModule recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+
+               /*
+                * At least, one clear scan (after the Transaction Manager has been shut down)
+                * should be completed to make sure that there are no transactions to recover.
+                */
+               doScanningWait();
+               doWork();
+
+               /*
+                * Now, it is finally possible to start checking if there are transactions that
+                * are still in need of recovery (or completion, in case of heuristics).
+                * The Action Manager gets checked as in-flight transactions with timeout zero
+                * can still be around (as the Transaction Reaper doesn't handle those txns)
+                */
+               while (ActionManager.manager().getNumberOfInflightTransactions() != 0 || _blockSuspension) {
```

Comment:
Actually sorry, do you think this waiting for `ActionManager.manager().getNumberOfInflightTransactions() != 0` should be part of the TransactionManager suspend call instead - I think it maybe should be as it is more about the suspension of the transaction manager IIUC?

Reply:
Great observation. While I was adding this check here, I was thinking that I should find a better place for it. I will think more about your comment. Thanks

Reply:
Please consider this resolved